cmd/data.go

```diff
@@ -112,77 +114,110 @@ func (cmd *DataCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface
+		utils.PrintPermissionsWarning(sourceProfile.Driver, ioHelper.Out)
 
-	if !sourceProfile.UseTargetSchema() {
-		err = conversion.CreateOrUpdateDatabase(ctx, adminClient, dbURI, sourceProfile.Driver, targetProfile.TargetDb, conv, ioHelper.Out)
+		dbURI := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName)
+		client, err := utils.GetClient(ctx, dbURI)
 		if err != nil {
-			err = fmt.Errorf("can't create/update database: %v", err)
+			fmt.Printf("can't create client for db %s: %v\n", dbURI, err)
 			return subcommands.ExitFailure
 		}
-	}
+		defer client.Close()
 
-	dataCoversionStartTime := time.Now()
-	bw, err := conversion.DataConv(ctx, sourceProfile, targetProfile, &ioHelper, client, conv, true, cmd.writeLimit)
-	if err != nil {
-		err = fmt.Errorf("can't finish data migration: %v", err)
-		return subcommands.ExitFailure
-	}
-	dataCoversionEndTime := time.Now()
-	dataCoversionDuration := dataCoversionEndTime.Sub(dataCoversionStartTime)
-	conv.Audit.DataConversionDuration = dataCoversionDuration
+		adminClient, err := utils.NewDatabaseAdminClient(ctx)
+		if err != nil {
+			fmt.Printf("can't create admin client: %v\n", utils.AnalyzeError(err, dbURI))
+			return subcommands.ExitFailure
+		}
+		defer adminClient.Close()
+		if !sourceProfile.UseTargetSchema() {
+			dbExists, err := conversion.CheckExistingDb(ctx, adminClient, dbURI)
+			if err != nil {
+				fmt.Printf("can't verify target database: %v\n", err)
+				return subcommands.ExitFailure
+			}
+			if !dbExists {
+				fmt.Println("target database doesn't exist")
+				return subcommands.ExitFailure
+			}
+			err = conversion.ValidateTables(ctx, client, conv.TargetDb)
+			if err != nil {
+				fmt.Printf("can't validate the tables: %v\n", err)
+				return subcommands.ExitFailure
+			}
+			spannerConv := internal.MakeConv()
+			spannerConv.TargetDb = targetProfile.TargetDb
+			err = utils.ReadSpannerSchema(ctx, spannerConv, client)
+			if err != nil {
+				fmt.Printf("can't read spanner schema: %v\n", err)
+				return subcommands.ExitFailure
+			}
+			err = utils.CompareSchema(conv, spannerConv)
+			if err != nil {
+				fmt.Printf("error while comparing the schema from session file and existing spanner schema: %v\n", err)
+				return subcommands.ExitFailure
+			}
+		}
+		bw, err = conversion.DataConv(ctx, sourceProfile, targetProfile, &ioHelper, client, conv, true, cmd.writeLimit)
+		if err != nil {
+			fmt.Printf("can't finish data conversion for db %s: %v\n", dbURI, err)
+			return subcommands.ExitFailure
+		}
 
-	if !cmd.skipForeignKeys {
 		if err = conversion.UpdateDDLForeignKeys(ctx, adminClient, dbURI, conv, ioHelper.Out); err != nil {
-			err = fmt.Errorf("can't perform update schema on db %s with foreign keys: %v", dbURI, err)
+			fmt.Printf("can't perform update schema on db %s with foreign keys: %v\n", dbURI, err)
+			return subcommands.ExitFailure
+		}
+		banner = utils.GetBanner(now, dbURI)
+	} else {
```

Comment:
Refactoring: instead of formatting it like
 ```
if dryRun {
  do abc
} else {
  do xyz
}
```
it will be much more readable with something like

```
if dryRun
   funcToDoabc()
else 
   funcToDoxyz()
```
And I think we should be able to reuse funcToDoabc90 and xyz() in all 3 cmd files

Reply:
Used functions in case of repeated code and modified the PR